### PR TITLE
Add shading to Axes3D.voxels, and enable it by default

### PR DESCRIPTION
**Pull request by eric-wieser**
## PR Summary

The conclusion of my normals and shading work (#12792, #12136 ).

The rich image diffs in the patch should show the effect. For a quick summary, here's what the output now looks like:

## PR Checklist

- [ ] Has Pytest style unit tests
  - Do the existing ones count? Should I set `shade=False` on one of the existing tests, or parametrize them all on shaded vs not
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
  - Awaiting CI
- [x] New features are documented, with examples if plot related
  - Do existing examples count?
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
   - Is this major enough?
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way
  - Not backward-incompatible

### CODE REVIEW

**Comment by timhoffm** [comment]
> I assume pytest failures are related to #13122

Yes, you can rebase now that this is in.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -2872,7 +2884,8 @@ def permutation_matrices(n):
             qinds = np.arange(qc)
             rinds = np.arange(rc)
 
-            square_rot = square.dot(permute.T)
+            square_rot_pos = square.dot(permute.T)
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py]:
```suggestion
            square_rot_pos = square @ permute.T
```
We can use the matrix multiplication operator now that we are on Py3.

**Comment by eric-wieser** [comment]
Thanks @timhoffm, done

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (eric-wieser) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Not a bad idea - although your suggested change doesn't let me change both at once through the web.

Has matplotlib already switched all uses of `.dot` to `@` internally? If not, I'd be inclined to declare this out of scope, and leave it for later cleanup.

Comment (timhoffm) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: No, we have not. You may leave it for later if you wish to.

Comment (eric-wieser) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Will change if another revision is justified for another reason.

**Comment by timhoffm** [comment]
flake8 complains

~~~
./lib/mpl_toolkits/mplot3d/axes3d.py:2947:80: E501 line too long (83 > 79 characters)
1     E501 line too long (83 > 79 characters)
~~~

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (eric-wieser) [reply lib/mpl_toolkits/mplot3d/axes3d.py]: Actually, I'd like to leave that to an axes3d-wide cleanup of `dot`

**Comment by NelleV** [comment]
This is really cool!
It needs a "what's new" entry (in users/next_whats_new/)
I also think we should update one of the examples from our gallery with this option. How about this one: https://matplotlib.org/devdocs/gallery/mplot3d/voxels.html#sphx-glr-gallery-mplot3d-voxels-py ?

**Comment by eric-wieser** [comment]
Shading is now the default - all of those gallery examples will use it now.

I'll add the what's new when I get around to it.

**Comment by NelleV** [comment]
Sorry, I missed that. 

That's a backward incompatible change on the design. @tacaswell do we still care about those?

**Comment by eric-wieser** [comment]
Yep, it is. That's the main thing I wanted feedback on

**Comment by eric-wieser** [comment]
How was shading introduced for the other functions? Or was it always there for bar3d, trisurf, ...

**Comment by WeatherGod** [comment]
it was always there

**Comment by anntzer** [comment]
Milestoning as 3.1 so that we can at least reach a decision of whether to change the default there or not (the longer we wait, the less reasonable it is to change the default without a deprecation period).

**Comment by ImportanceOfBeingErnest** [comment]
Concerning changing the defaults: With https://github.com/matplotlib/matplotlib/pull/12259 being merged, all 3d bar plots will now have a different shading than before. I don't think #12259 has any "what's new" or similar in it, and it also will break someone's image comparisson.

**Comment by efiring** [comment]
I'm convinced by the argument that it is OK to proceed with this change, since voxels are quite new, etc. ping @dstansby

**Comment by eric-wieser** [comment]
I'll try to add a whats-new entry today.

**Comment by eric-wieser** [comment]
What's new entry added

**Comment by eric-wieser** [comment]
[Screwed up the plot a little](https://18433-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/users/next_whats_new/2019-01-06-shaded-voxels.html), trying again

**Comment by eric-wieser** [comment]
Fixed plot: https://18441-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/users/next_whats_new/2019-01-06-shaded-voxels.html

**Comment by tacaswell** [comment]
I think this is ok to change; it is a pretty clear improvement to a new feature.

I added a file to the `next_api_changes` so it will be noted there as well, anyone can merge once CI finishes again (just be paranoid about sphinx).

**Pull request merged**